src/rslice.rs

```diff
@@ -0,0 +1,188 @@
+}
+
+/// NulByte represents an error converting `&str` to `rustls_str` when the &str
+/// contains a NUL.
+type NulByte = ();
+
+impl<'a> TryFrom<&'a str> for rustls_str<'a> {
+    type Error = NulByte;
+
+    fn try_from(s: &str) -> Result<Self, Self::Error> {
+        if s.contains('\0') {
+            return Err(());
+        }
+        Ok(rustls_str {
+            data: s.as_ptr() as *const c_char,
+            len: s.len() as size_t,
+            phantom: PhantomData,
+        })
+    }
+}
+
+/// An owned Vec<rustls_str>, where the inner slices have lifetime 'a.
+/// See comment on VecSliceBytes for more information.
+pub(crate) struct VecStr<'a>(Vec<rustls_str<'a>>);
+
+impl<'a> VecStr<'a> {
+    /// Build a VecStr that refers to `input` and can live as long as it does.
+    fn new(input: &'a Vec<&str>) -> Result<Self, ()> {
+        let mut vs: Vec<rustls_str> = vec![];
+        for v in input {
+            let v: &str = v.as_ref();
+            vs.push(v.try_into()?);
+        }
+        Ok(VecStr(vs))
+    }
+}
+
+/// A read-only view of a slice of Rust `&str`
+///
+/// This is used to pass data from crustls to callback functions provided
+/// by the user of the API. The `data` is an array of `rustls_str`
+/// structures with `len` elements.
+///
+/// The memory exposed is available as specified by the function
+/// using this in its signature. For instance, when this is a parameter to a
+/// callback, the lifetime will usually be the duration of the callback.
+/// Functions that receive one of these must not dereference any of the
+/// involved pointers beyond the allowed lifetime.
+#[repr(C)]
+pub struct rustls_slice_str<'a> {
+    pub data: *const rustls_str<'a>,
+    pub len: size_t,
+    phantom: PhantomData<&'a [rustls_str<'a>]>,
+}
+
+impl<'a> From<&'a VecStr<'a>> for rustls_slice_str<'a> {
+    fn from(input: &'a VecStr<'a>) -> Self {
+        rustls_slice_str {
+            data: input.0.as_ptr(),
+            len: input.0.len(),
+            phantom: PhantomData,
+        }
+    }
+}
+
+/// A read-only view on a Rust slice of 16-bit integers in platform endianness.
+///
+/// This is used to pass data from crustls to callback functions provided
+/// by the user of the API.
+/// `len` indicates the number of bytes than can be safely read.
+///
+/// The memory exposed is available as specified by the function
+/// using this in its signature. For instance, when this is a parameter to a
+/// callback, the lifetime will usually be the duration of the callback.
+/// Functions that receive one of these must not dereference the data pointer
+/// beyond the allowed lifetime.
+#[repr(C)]
+pub struct rustls_slice_u16<'a> {
+    pub data: *const u16,
+    pub len: size_t,
+    phantom: PhantomData<&'a [u16]>,
+}
+
+impl<'a> From<&'a [u16]> for rustls_slice_u16<'a> {
+    fn from(s: &[u16]) -> Self {
+        rustls_slice_u16 {
+            data: s.as_ptr() as *const u16,
+            len: s.len() as size_t,
```

Comment:
FYI, both these casts are redundant, and the from/new functions aren't consistent on whether they do or don't cast.